Tactics/CSE.lean

```diff
@@ -218,56 +272,81 @@ partial def CSEM.planCSE (e : Expr): CSEM GeneralizeArg := do
 Plan to perform a CSE for this expression, by building a 'GeneralizeArg'.
 -/
 def CSEM.generalize (arg : GeneralizeArg) : CSEM Bool := do
+  consumeFuelEliminate
+  unless (← hasFuelEliminate?) do
+    trace[Tactics.cse.summary] "⏸️ CSE ran out of fuel while performing subexpression elimination. increase `fuelEliminate`."
+    return false
+
   let hname := arg.hName?.getD .anonymous
   let xname := arg.xName?.getD .anonymous
   let e := arg.expr
 
   let mvarId ← getMainGoal
   mvarId.withContext do
     -- implementation modeled after `Lean.Elab.Tactic.evalGeneralize`.
-    trace[Tactic.cse.generalize] "{tryEmoji} Generalizing {hname} : {e} = {xname}"
+    traceLargeMsg m!"{tryEmoji} Generalizing {hname}: {xname} = ..."  m!"{e}"
+    let checkpoint ← Tactic.saveState
     try
-      -- Implementation modeled after `Lean.MVarId.generalizeHyp`.
-      let e ← instantiateMVars e
-      let hyps := ((← getLCtx).getFVarIds)
-      let transparency := TransparencyMode.reducible
-      let hyps ← hyps.filterM fun h => do
-        let type ← instantiateMVars (← h.getType)
-        return (← withTransparency transparency <| kabstract type arg.expr).hasLooseBVars
-      let (reverted, mvarId) ← mvarId.revert hyps true
-      let (_newVars, mvarId) ← mvarId.generalize #[arg] transparency
-      let (reintros, mvarId) ← mvarId.introNP reverted.size
-      let _fvarSubst := Id.run do
-        let mut subst : FVarSubst := {}
-        for h in reverted, reintro in reintros do
-          subst := subst.insert h (mkFVar reintro)
-        pure subst
-      replaceMainGoal [mvarId]
-      trace[Tactic.cse.generalize] "{checkEmoji} succeeded in generalizing {hname}. ({← getMainGoal})"
+      if ! (← isDryRun?) then
+        -- Implementation modeled after `Lean.MVarId.generalizeHyp`.
+        let e ← instantiateMVars e
+        let hyps := ((← getLCtx).getFVarIds)
+        let transparency := TransparencyMode.reducible
+        let hyps ← hyps.filterM fun h => do
+          let type ← instantiateMVars (← h.getType)
+          return (← withTransparency transparency <| kabstract type arg.expr).hasLooseBVars
+        let (reverted, mvarId) ← mvarId.revert hyps true
+        let (_newVars, mvarId) ← mvarId.generalize #[arg] transparency
+        let (reintros, mvarId) ← mvarId.introNP reverted.size
+        let _fvarSubst := Id.run do
+          let mut subst : FVarSubst := {}
+          for h in reverted, reintro in reintros do
+            subst := subst.insert h (mkFVar reintro)
+          pure subst
+        replaceMainGoal [mvarId]
+      traceLargeMsg m!"{checkEmoji} succeeded in generalizing {hname}." m!"{← getMainGoal}"
       return true
     catch e =>
-      trace[Tactic.cse.generalize] "{bombEmoji} failed to generalize {hname}"
+      checkpoint.restore
+      traceLargeMsg m!"{bombEmoji} failed to generalize {hname}" m!"{e.toMessageData}"
```

Comment:
The bug was because we didn't restore state correctly -- fixing this fixed the problem of "no goals to be solved".